src/output-filestore.c

```diff
@@ -0,0 +1,482 @@
+{
+    int i = 0;
+    size_t x = 0;
+    for (i = 0, x = 0; x < buf_len; x++) {
+        i += snprintf(&str[i], size - i, "%02x", buf[x]);
+    }
+}
+
+/**
+ * \brief Update the timestamps on a file to match those of another
+ *     file.
+ *
+ * \param src_filename Filename to use as timestamp source.
+ * \param filename Filename to apply timestamps to.
+ */
+static void OutputFilestoreUpdateFileTime(const char *src_filename,
+        const char *filename)
+{
+    struct stat sb;
+    if (stat(src_filename, &sb) != 0) {
+        SCLogDebug("Failed to stat %s: %s", filename, strerror(errno));
+        return;
+    }
+    struct utimbuf utimbuf = {
+        .actime = sb.st_atime,
+        .modtime = sb.st_mtime,
+    };
+    if (utime(filename, &utimbuf) != 0) {
+        SCLogDebug("Failed to update file timestamps: %s: %s", filename,
+                strerror(errno));
+    }
+}
+
+static void OutputFilestoreFinalizeFiles(const OutputFilestoreCtx *ctx,
+        const Packet *p, File *ff) {
+    /* Stringify the SHA256 which will be used in the final
+     * filename. */
+    char sha256string[(SHA256_LENGTH * 2) + 1];
+    PrintHexString(sha256string, sizeof(sha256string), ff->sha256,
+            sizeof(ff->sha256));
+
+    char tmp_filename[PATH_MAX] = "";
+    snprintf(tmp_filename, sizeof(tmp_filename), "%s/file.%u", ctx->tmpdir,
+            ff->file_store_id);
+
+    char final_filename[PATH_MAX] = "";
+    snprintf(final_filename, sizeof(final_filename), "%s/%c%c/%s",
+            ctx->prefix, sha256string[0], sha256string[1], sha256string);
+
+    if (SCPathExists(final_filename)) {
+        OutputFilestoreUpdateFileTime(tmp_filename, final_filename);
+        if (unlink(tmp_filename) != 0) {
+            SCLogWarning(SC_WARN_REMOVE_FILE,
+                    "Failed to remove temporary file %s: %s", tmp_filename,
+                    strerror(errno));
+        }
+    } else if (rename(tmp_filename, final_filename) != 0) {
+        SCLogWarning(SC_WARN_RENAMING_FILE, "Failed to rename %s to %s: %s",
+                tmp_filename, final_filename, strerror(errno));
+        return;
+    }
+
+#ifdef HAVE_LIBJANSSON
+    if (ctx->fileinfo) {
+        char js_metadata_filename[PATH_MAX];
+        snprintf(js_metadata_filename, sizeof(js_metadata_filename),
+                "%s.%lu.%d.json", final_filename, p->ts.tv_sec,
+                ff->file_store_id);
+        json_t *js_fileinfo = JsonBuildFileInfoRecord(p, ff);
+        json_dump_file(js_fileinfo, js_metadata_filename, 0);
+        json_decref(js_fileinfo);
+    }
+#endif
+}
+
+static int OutputFilestoreLogger(ThreadVars *tv, void *thread_data,
+        const Packet *p, File *ff, const uint8_t *data, uint32_t data_len,
+        uint8_t flags)
+{
+    SCEnter();
+    OutputFilestoreLogThread *aft = (OutputFilestoreLogThread *)thread_data;
+    OutputFilestoreCtx *ctx = aft->ctx;
+    char filename[PATH_MAX] = "";
+    int file_fd = -1;
+
+    /* no flow, no htp state */
```

Comment:
outdated comment, 'no flow, no files' would be more accurate

Reply:
Fixed.